File: models/claims/claim.js
const { DataTypes } = require('sequelize');
const sequelize = require('../../config/database');
const Visit = require('../Visit');
const ClaimItem = require('./claimItem');


const Claim = sequelize.define('Claim', {
    id: {
        type: DataTypes.UUID,
        primaryKey: true,
        defaultValue: DataTypes.UUIDV4
    },
    visit_id: {
        type: DataTypes.UUID,
        allowNull: false
    },
    item_id: {
        type: DataTypes.UUID,
        allowNull: true
    },
    claim_status: {
        type: DataTypes.ENUM('Pending', 'Submitted', 'Approved', 'Rejected', 'Resubmitted', 'Draft'),
        defaultValue: 'Pending'
    },
    submission_date: {
        type: DataTypes.DATE
    },
    total_amount: {
        type: DataTypes.FLOAT,
        defaultValue: 0
    },
    total_nhia_amount: {
        type: DataTypes.FLOAT,
        allowNull: true,
        defaultValue: 0,
        comment: 'Sum of NHIA-covered amounts across all claim items'
    },
    total_patient_amount: {
        type: DataTypes.FLOAT,
        allowNull: true,
        defaultValue: 0,
        comment: 'Sum of patient co-payments across all claim items'
    },
    claim_reference_number: {
        type: DataTypes.STRING,
        allowNull: false,
        // unique: true,
        // comment: 'Unique reference number for the claim'
    },
    // Add to your Claim model
    batch_id: {
        type: DataTypes.UUID,
        allowNull: true,
        defaultValue: DataTypes.UUIDV4
    }

}, {
    timestamps: true,
    tableName: 'claims',
    comment: 'Table to store claims information',
});


Claim.associate = (models) => {
    Claim.belongsTo(models.Visit, {
        foreignKey: 'visit_id',
        as: 'visit'
    });

    Claim.hasMany(models.ClaimItem, {
        foreignKey: 'claim_id',
        as: 'items'
    });
};




module.exports = Claim;

File: claims-service/app/models/claim.py
"""Claim model - NHIA claim linked to a patient visit."""

import uuid
from datetime import datetime

from sqlalchemy import Column, DateTime, Float, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.database import Base


class Claim(Base):
    __tablename__ = "claims"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    visit_id = Column(UUID(as_uuid=True), nullable=False, index=True)
    item_id = Column(UUID(as_uuid=True), nullable=True)

    # Use String instead of Enum to avoid mismatch between Python enum values
    # and PostgreSQL enum labels. The DB stores 'Pending', 'Draft', etc.
    claim_status = Column(String(50), default="Pending", index=True)

    submission_date = Column(DateTime(timezone=True), nullable=True)
    total_amount = Column(Float, default=0)
    total_nhia_amount = Column(Float, default=0, comment="Sum of NHIA-covered amounts")
    total_patient_amount = Column(Float, default=0, comment="Sum of patient co-payments")
    claim_reference_number = Column(String(255), nullable=False, index=True)
    batch_id = Column(UUID(as_uuid=True), nullable=True, index=True)

    # Map to camelCase DB columns (Sequelize convention)
    created_at = Column("createdAt", DateTime(timezone=True), default=datetime.utcnow)
    updated_at = Column("updatedAt", DateTime(timezone=True), default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    items = relationship("ClaimItem", back_populates="claim", lazy="selectin")

    def recompute_totals(self):
        """Recalculate totals from claim items."""
        self.total_amount = sum(float(i.amount or 0) for i in self.items)
        self.total_nhia_amount = sum(float(i.nhia_amount or 0) for i in self.items)
        self.total_patient_amount = sum(float(i.co_payment or 0) for i in self.items)
